kernel/apic.c

```diff
@@ -57,11 +158,99 @@ static uint8_t analyze_madt_entry (uint8_t *entry) {
+}
+
+void apic_eoi (void) {
+  apic->end_of_interrupt.val = 0;
+}
+
+/* Format for APIC ICR registers and for IOAPIC redirection table registers. */
+
+static const uint32_t APIC_ICR_LOW_DELIVERY_MODE_FIXED          = 0x00000000;
+static const uint32_t APIC_ICR_LOW_DELIVERY_MODE_LOW_PRIORITY   = 0x00000100;
+static const uint32_t APIC_ICR_LOW_DELIVERY_MODE_SMI            = 0x00000200;
+static const uint32_t APIC_ICR_LOW_DELIVERY_MODE_NMI            = 0x00000400;
+static const uint32_t APIC_ICR_LOW_DELIVERY_MODE_INIT           = 0x00000500;
+static const uint32_t APIC_ICR_LOW_DELIVERY_MODE_START_UP       = 0x00000600;
+
+static const uint32_t APIC_ICR_LOW_DESTINATION_MODE_LOGICAL     = 0x00000800;
+
+static const uint32_t APIC_ICR_LOW_DELIVERY_STATUS_SEND_PENDING = 0x00001000;
+
+static const uint32_t APIC_ICR_LOW_PIN_POLARITY_ACTIVE_LOW      = 0x00002000;
+
+static const uint32_t APIC_ICR_LOW_LEVEL_ASSERT                 = 0x00004000;
+
+static const uint32_t APIC_ICR_LOW_TRIGGER_LEVEL                = 0x00008000;
+
+static const uint32_t APIC_ICR_LOW_MASK_LINE                    = 0x00010000;
+
+static const uint32_t APIC_ICR_LOW_DEST_SHORTHAND_SELF          = 0x00040000;
+static const uint32_t APIC_ICR_LOW_DEST_SHORTHAND_ALL_AND_SELF  = 0x00080000;
+static const uint32_t APIC_ICR_LOW_DEST_SHORTHAND_ALL_OTHERS    = 0x000C0000;
+
+static inline uint32_t APIC_ICR_HIGH_DESTINATION(uint32_t dest) {
+  return dest << 24;
+}
+
+static void write_ioapic_redirection (uint32_t line, uint32_t low_flags, uint32_t high_flags) {
+  ioapic->reg = IOAPIC_RED_TBL_LOW(line);
+  ioapic->data = low_flags;
+  ioapic->reg = IOAPIC_RED_TBL_HIGH(line);
+  ioapic->data = high_flags;
+}
+
+static void map_line (uint8_t line) {
+  /* Map ISA line n to n+64 unless there is a remapping. */
+  for (uint8_t i = 0; i < nremaps; i++) {
+    if (remaps[i].isa_line == line) {
+      uint32_t low_flags = (line + 64)
+        | (remaps[i].polarity_is_low ? APIC_ICR_LOW_PIN_POLARITY_ACTIVE_LOW : 0)
+        | (remaps[i].trigger_is_level ? APIC_ICR_LOW_TRIGGER_LEVEL : 0)
+        | APIC_ICR_LOW_MASK_LINE;
+      uint32_t high_flags = APIC_ICR_HIGH_DESTINATION(0);
+      write_ioapic_redirection(remaps[i].remapped_apic_line, low_flags, high_flags);
+      return;
+    }
+  }
+  for (uint8_t i = 0; i < nremaps; i++) {
+    if (remaps[i].remapped_apic_line == line) {
+      return;
+    }
+  }
+  uint32_t low_flags = (line + 64) | APIC_ICR_LOW_MASK_LINE;
+  uint32_t high_flags = APIC_ICR_HIGH_DESTINATION(0);
+  write_ioapic_redirection(line, low_flags, high_flags);
+}
+
+void ioapic_mask_line (uint32_t line, uint8_t masked) {
+  ioapic->reg = IOAPIC_RED_TBL_LOW(line);
+  if (masked) {
+    ioapic->data |= APIC_ICR_LOW_MASK_LINE;
+  } else {
+    ioapic->data &= ~APIC_ICR_LOW_MASK_LINE;
+  }
+}
+
+void ioapic_init () {
+  for (uint8_t line = 0; line < 16; line++) {
+    map_line(line);
+  }
+  ioapic_mask_line(0x02, 0);
```

Comment:
what are you masking?

Reply:
Nothing, it turns out.  I added a CL that fulfills the TODO in this code, and also refactors the API exposed outside of apic.h to be a little friendlier.